flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/factory/KubernetesFactoryWithPodTemplateTestBase.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.kubeclient.factory;
+
+import org.apache.flink.configuration.BlobServerOptions;
+import org.apache.flink.configuration.MemorySize;
+import org.apache.flink.configuration.RestOptions;
+import org.apache.flink.configuration.TaskManagerOptions;
+import org.apache.flink.kubernetes.KubernetesPodTemplateTestUtils;
+import org.apache.flink.kubernetes.KubernetesTestBase;
+import org.apache.flink.kubernetes.configuration.KubernetesConfigOptionsInternal;
+import org.apache.flink.kubernetes.entrypoint.KubernetesSessionClusterEntrypoint;
+import org.apache.flink.kubernetes.utils.Constants;
+
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.PodSpec;
+import org.junit.Test;
+
+import static org.hamcrest.Matchers.hasItems;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Test base for the {@link KubernetesJobManagerFactory} and {@link KubernetesTaskManagerFactory}
+ * with pod template. These tests will ensure that annotations, labels, imagePullSecrets,
+ * nodeSelector, tolerations, env, init container, sidecar container, volumes from pod template
+ * should be kept after all decorators.
+ */
+public abstract class KubernetesFactoryWithPodTemplateTestBase extends KubernetesTestBase {
```

Comment:
In general I would be cautious with using inheritance for test classes. This makes it a lot harder for someone new to this part of the code to understand what is actually happening. In most cases, one can avoid test bases by using composition (e.g. creating a `KubernetesTestingResource` which sets up the main part and can be used by test clases which want to test something for K8s).

Reply:
The only reason why I introduce this abstract class is to avoid duplicated tests for `KubernetesJobManagerFactoryWithPodTemplateTest` and `KubernetesTaskManagerFactoryWithPodTemplateTest`. It is not about setting up the main part.

The implementation classes need to `setResultPod` and then the result pod will be verified in the base class.